src/services/location.js

```diff
@@ -47,48 +45,129 @@ ngeo.Location = function(location, history) {
- * @export
- */
-ngeo.Location.prototype.getUri = function() {
-  return this.uri_;
+  /**
+   * @type {number|undefined}
+   * @private
+   */
+  this.port_ = location.port ? parseInt(location.port, 10) : undefined;
+
+  /**
+   * @type {string|undefined}
+   * @private
+   */
+  this.path_ = location.pathname;
+
+  function fillQueryData(queryData, encodedQuery) {
+    if (encodedQuery) {
+      const pairs = encodedQuery.split('&');
+      for (const pair of pairs) {
+        const indexOfEquals = pair.indexOf('=');
+        if (indexOfEquals >= 0) {
+          const name = pair.substring(0, indexOfEquals);
+          const value = pair.substring(indexOfEquals + 1);
+          queryData[decodeURIComponent(name)] = decodeURIComponent(value);
+        } else {
+          queryData[pair] = '';
+        }
+      }
+    }
+  }
+
+  /**
+   * @type {!Object.<string, string>}
+   * @private
+   */
+  this.queryData_ = {};
+  if (location.search) {
+    fillQueryData(this.queryData_, location.search.substring(1));
+  }
+
+  /**
+   * @type {!Object.<string, string>}
+   * @private
+   */
+  this.fragment_ = {};
+  if (location.hash) {
+    fillQueryData(this.fragment_, location.hash.substring(1));
+  }
 };
 
 
 /**
  * Get the location's current path.
- * @return {string} The path.
+ * @return {string|undefined} The path.
  * @export
  */
 ngeo.Location.prototype.getPath = function() {
-  return this.uri_.getPath();
+  return this.path_;
 };
 
 
 /**
  * Get the location's URI as a string
- * @param {Object.<string, string>=} opt_params Params.
  * @return {string} The URI.
  * @export
  */
-ngeo.Location.prototype.getUriString = function(opt_params) {
-  let extendedUri;
-  if (opt_params !== undefined) {
-    extendedUri = this.uri_.clone();
-    extendedUri.getQueryData().extend(opt_params);
-  } else {
-    extendedUri = this.uri_;
+ngeo.Location.prototype.getUriString = function() {
+  const out = [];
+
+  if (this.schema_) {
+    out.push(this.schema_, ':');
+  }
+
+  if (this.domain_ || this.schema_ == 'file') {
+    out.push('//');
+
+    out.push(this.domain_);
+
+    if (this.port_ !== undefined) {
+      out.push(':', String(this.port_));
+    }
+  }
+
+  if (this.path_) {
+    if (this.domain_ && this.path_.charAt(0) != '/') {
+      out.push('/');
+    }
+    out.push(this.path_);
+  }
+
+  function getEncodedQuery(queryData) {
+    const sb = [];
+    // In the past, we use this.getKeys() and this.getVals(), but that
+    // generates a lot of allocations as compared to simply iterating
+    // over the keys.
+    for (const key in queryData) {
+      const value = queryData[key];
+      if (value) {
+        sb.push(`${encodeURIComponent(key)}=${encodeURIComponent(value)}`);
+      } else {
+        sb.push(encodeURIComponent(key));
+      }
+    }
+    return sb;
+  }
+  const eqs = getEncodedQuery(this.queryData_);
+  if (eqs.length > 0) {
+    out.push('?', eqs.join('&'));
+  }
+
+  const ef = getEncodedQuery(this.fragment_);
```

Comment:
For the second, I don't totally agree because we use the standard query syntax for the fragment and that's not standard...